hal/src/rtl872x/interrupts_hal.cpp

```diff
@@ -48,12 +53,87 @@ typedef enum interrupt_state_t {
     INT_STATE_SUSPENDED
 } interrupt_state_t;
 
+#if MODULE_FUNCTION != MOD_FUNC_BOOTLOADER
+
+class InterruptConfig {
+public:
+    struct InterruptCallback {
+        bool operator==(const InterruptCallback& callback) {
+            return handler == callback.handler;
+        }
+
+        hal_interrupt_handler_t handler;
+        void* data;
+        uint32_t priority; // The lower the value, the higher the priority
+    };
+
+    InterruptConfig()
+            : state(INT_STATE_DISABLED),
+              callbacks_() {
+    }
+    ~InterruptConfig() = default;
+
+    int appendHandler(hal_interrupt_handler_t handler, void* data, uint32_t priority) {
+        uint32_t idx = 0;
+        InterruptCallback cb = {};
+        cb.handler = handler;
+        cb.data = data;
+        cb.priority = priority;
+        for (auto& callback : callbacks_) {
+            if (priority < callback.priority) {
+                break;
+            }
+            idx++;
+        }
+        CHECK_TRUE(callbacks_.insert(idx, cb), SYSTEM_ERROR_NO_MEMORY);
+        return SYSTEM_ERROR_NONE;
+    }
+
+    ssize_t handlers() const {
+        return callbacks_.size();
+    }
+
+    int removeHandler(hal_interrupt_handler_t handler) {
+        for (auto& callback : callbacks_) {
+            if (handler == callback.handler) {
+                callbacks_.removeOne(callback);
+            }
+        }
+        return SYSTEM_ERROR_NONE;
+    }
+
+    void clearHandlers() {
+        callbacks_.removeAt(0, callbacks_.size());
+    }
+
+    void executeHandlers() const {
+        for (auto& callback : callbacks_) {
+            if (callback.handler) {
+                callback.handler(callback.data);
+            }
+        }
+    }
+
+public:
+    interrupt_state_t           state;
+    InterruptMode               mode;
+
+private:
+    Vector<InterruptCallback>   callbacks_;
```

Comment:
There is no thread safety in interrupts hal and with existing implementation it all works out okay, but adding a Vector here is going to cause thread/interrupt-safety issues when modifying handlers/trying to execute them.

So, we either have to add some form of interrupt/thread-safety e.g. create a new modified vector of handlers and swap it with existing one when modifying while disabling interrupts.

Reply:
Can we simply disable interrupt while modifying the verctor?

Reply:
No, we unfortunately cannot, as modifying a vector involves heap allocations and those cannot be done under disabled interrupts. You could use an `IntrusiveList` perhaps and perform allocations outside of the list pointer modifications as an option if we are to step away from a vector here.

Reply:
You can manually change `next` pointer though. `AtomicSimpleStaticPool` is an overkill for this honestly, better create a whole new vector and swap atomically in that case then.